File: webshare_app/security/permissions.py
"""
WebShare Pro - Folder permissions.
"""

from __future__ import annotations

import json
import os
from typing import Any

from config import FOLDER_PERMISSIONS, PERMISSIONS_FILE, conf, permissions_lock
from utils.file_utils import validate_path
from utils.log_manager import logger
from webshare_app.core.persistence import persist_json_snapshot

PERMISSIONS_PERSIST_KEY = "folder_permissions"

VALID_PERMISSION_ACTIONS = {"read", "write", "delete"}
VALID_PERMISSION_PRINCIPALS = {"admin", "guest", "*"}
DEFAULT_PERMISSION = {"read": ["*"], "write": ["*"], "delete": ["admin"]}


def _is_protected_permission_path(path: str) -> bool:
    for segment in path.split("/"):
        if not segment:
            continue
        lower = segment.lower()
        if segment.startswith(".") or lower.startswith(".webshare"):
            return True
    return False


def normalize_permission_path(path: str | None) -> str:
    value = str(path or "").replace("\\", "/").strip("/")
    parts: list[str] = []
    for part in value.split("/"):
        part = part.strip()
        if not part or part == ".":
            continue
        if part == "..":
            raise ValueError("invalid permission path")
        parts.append(part)

    normalized = "/".join(parts)
    if not normalized or _is_protected_permission_path(normalized):
        raise ValueError("invalid permission path")

    valid, _full_path, _error = validate_path(conf.get("folder"), normalized)
    if not valid:
        raise ValueError("invalid permission path")
    return normalized


def normalize_permission_users(users: Any) -> list[str]:
    if not isinstance(users, list):
        raise ValueError("invalid permission users")
    normalized: list[str] = []
    for user in users:
        value = str(user or "").strip()
        if value not in VALID_PERMISSION_PRINCIPALS:
            raise ValueError("invalid permission user")
        if value not in normalized:
            normalized.append(value)
    return normalized


def normalize_permission_entry(path: str | None, data: dict[str, Any]) -> tuple[str, dict[str, list[str]]]:
    if not isinstance(data, dict):
        raise ValueError("invalid permission entry")

    normalized_path = normalize_permission_path(path)
    entry: dict[str, list[str]] = {}
    for action, users in data.items():
        if action not in VALID_PERMISSION_ACTIONS:
            raise ValueError("invalid permission action")
        entry[action] = normalize_permission_users(users)
    if not entry:
        raise ValueError("empty permission entry")
    return normalized_path, entry


def check_permission(path: str, user: str, action: str) -> bool:
    """Check inherited read/write/delete permissions. Admin is always allowed."""
    if user == "admin":
        return True
    if action not in VALID_PERMISSION_ACTIONS:
        return False

    normalized = str(path or "").replace("\\", "/").strip("/")
    with permissions_lock:
        current_path = ""
        for part in normalized.split("/"):
            if not part:
                continue
            current_path = f"{current_path}/{part}" if current_path else part
            perm = FOLDER_PERMISSIONS.get(current_path)
            if not isinstance(perm, dict) or action not in perm:
                continue

            action_users = perm.get(action, [])
            if not isinstance(action_users, list):
                return False
            if "*" in action_users or user in action_users:
                continue
            return False

    return True


def _validated_permissions_snapshot() -> dict[str, dict[str, list[str]]]:
    snapshot: dict[str, dict[str, list[str]]] = {}
    for path, entry in FOLDER_PERMISSIONS.items():
        try:
            normalized_path, normalized_entry = normalize_permission_entry(path, entry)
        except ValueError:
            logger.add(f"Invalid permission skipped during save: {path}", "WARN")
            continue
        snapshot[normalized_path] = normalized_entry
    return snapshot


def save_permissions():
    """Persist folder permissions atomically."""
    base_dir = conf.get("folder")
    perm_path = os.path.join(base_dir, PERMISSIONS_FILE)
    try:
        persist_json_snapshot(
            PERMISSIONS_PERSIST_KEY,
            perm_path,
            permissions_lock,
            _validated_permissions_snapshot,
        )
    except Exception as exc:
        logger.add(f"permission save failed: {exc}", "ERROR")


def load_permissions():
    """Load folder permissions, skipping invalid legacy entries."""
    base_dir = conf.get("folder")
    perm_path = os.path.join(base_dir, PERMISSIONS_FILE)
    if not os.path.exists(perm_path):
        return

    try:
        with open(perm_path, "r", encoding="utf-8") as handle:
            loaded = json.load(handle)
    except Exception as exc:
        logger.add(f"permission load failed: {exc}", "ERROR")
        return

    normalized: dict[str, dict[str, list[str]]] = {}
    if isinstance(loaded, dict):
        for path, entry in loaded.items():
            try:
                normalized_path, normalized_entry = normalize_permission_entry(path, entry)
            except ValueError:
                logger.add(f"Invalid permission skipped: {path}", "WARN")
                continue
            normalized[normalized_path] = normalized_entry

    with permissions_lock:
        FOLDER_PERMISSIONS.clear()
        FOLDER_PERMISSIONS.update(normalized)
    logger.add(f"folder permissions loaded: {len(normalized)}")


def set_folder_permission(path: str, action: str, users: list):
    """Set a validated folder permission."""
    normalized_path = normalize_permission_path(path)
    if action not in VALID_PERMISSION_ACTIONS:
        raise ValueError("invalid permission action")
    normalized_users = normalize_permission_users(users)

    with permissions_lock:
        current = dict(DEFAULT_PERMISSION)
        current.update(FOLDER_PERMISSIONS.get(normalized_path, {}))
        current[action] = normalized_users
        FOLDER_PERMISSIONS[normalized_path] = current
    save_permissions()


def delete_folder_permission(path: str) -> bool:
    """Delete a validated folder permission."""
    try:
        normalized_path = normalize_permission_path(path)
    except ValueError:
        return False

    with permissions_lock:
        if normalized_path not in FOLDER_PERMISSIONS:
            return False
        del FOLDER_PERMISSIONS[normalized_path]
    save_permissions()
    logger.add(f"folder permission deleted: {normalized_path}")
    return True

File: go-core/internal/permission/permissions.go
package permission

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"strings"
	"sync"
)

// Folder-permission engine (parity with webshare_app/security/permissions.py
// and the mutation/capability helpers in utils/request_policy.py).
//
// Actions are exactly read/write/delete; principals are admin/guest/*.
// Admin bypasses all checks. Permissions inherit down the tree: walking from
// the root toward the target, any ancestor entry that names the action but
// does not grant the user denies (this matches Python's early-return).

// PermissionsFile is the store filename inside the shared folder.
const PermissionsFile = ".webshare_permissions.json"

// ValidActions and ValidPrincipals mirror the Python allow-lists.
var ValidActions = []string{"read", "write", "delete"}

// DefaultPermission mirrors DEFAULT_PERMISSION.
var DefaultPermission = map[string][]string{
	"read":   {"*"},
	"write":  {"*"},
	"delete": {"admin"},
}

// Capabilities mirrors build_path_capabilities exactly, including the
// parent-write rules for rename/upload/mkdir/unzip.
type Capabilities struct {
	Read   bool `json:"read"`
	Write  bool `json:"write"`
	Delete bool `json:"delete"`
	Rename bool `json:"rename"`
	Move   bool `json:"move"`
	Copy   bool `json:"copy"`
	Upload bool `json:"upload"`
	Mkdir  bool `json:"mkdir"`
	Edit   bool `json:"edit"`
	Trash  bool `json:"trash"`
	Unzip  bool `json:"unzip"`
}

func validAction(action string) bool {
	for _, a := range ValidActions {
		if a == action {
			return true
		}
	}
	return false
}

// NormalizePermissionPath mirrors normalize_permission_path: lexical
// cleanup, ".." rejected, empty/protected rejected, and the result must
// validate inside root.
func NormalizePermissionPath(root, path string) (string, error) {
	value := strings.ReplaceAll(path, "\\", "/")
	value = strings.Trim(value, "/")
	var parts []string
	for _, part := range strings.Split(value, "/") {
		part = strings.TrimSpace(part)
		if part == "" || part == "." {
			continue
		}
		if part == ".." {
			return "", errors.New("invalid permission path")
		}
		parts = append(parts, part)
	}
	normalized := strings.Join(parts, "/")
	if normalized == "" || IsProtectedSystemPath(normalized) {
		return "", errors.New("invalid permission path")
	}
	if ok, _, _ := ValidatePath(root, normalized); !ok {
		return "", errors.New("invalid permission path")
	}
	return normalized, nil
}

// NormalizePermissionUsers mirrors normalize_permission_users (dedup preserving order).
func NormalizePermissionUsers(users []string) ([]string, error) {
	var out []string
	seen := map[string]bool{}
	for _, u := range users {
		u = strings.TrimSpace(u)
		if u != "admin" && u != "guest" && u != "*" {
			return nil, errors.New("invalid permission user")
		}
		if !seen[u] {
			seen[u] = true
			out = append(out, u)
		}
	}
	return out, nil
}

// NormalizePermissionEntry mirrors normalize_permission_entry.
func NormalizePermissionEntry(root, path string, data map[string][]string) (string, map[string][]string, error) {
	if data == nil {
		return "", nil, errors.New("invalid permission entry")
	}
	normalizedPath, err := NormalizePermissionPath(root, path)
	if err != nil {
		return "", nil, err
	}
	entry := map[string][]string{}
	for action, users := range data {
		if !validAction(action) {
			return "", nil, errors.New("invalid permission action")
		}
		nu, err := NormalizePermissionUsers(users)
		if err != nil {
			return "", nil, err
		}
		entry[action] = nu
	}
	if len(entry) == 0 {
		return "", nil, errors.New("empty permission entry")
	}
	return normalizedPath, entry, nil
}

// Store is the in-memory folder-permission table with persistence.
type Store struct {
	mu      sync.RWMutex
	root    string
	entries map[string]map[string][]string
}

// NewStore creates a store rooted at the shared folder.
func NewStore(root string) *Store {
	return &Store{root: root, entries: map[string]map[string][]string{}}
}

// Root returns the shared folder root.
func (s *Store) Root() string { return s.root }

// Check mirrors check_permission (inherited, admin bypass).
func (s *Store) Check(path, user, action string) bool {
	if user == "admin" {
		return true
	}
	if !validAction(action) {
		return false
	}
	normalized := strings.Trim(strings.ReplaceAll(path, "\\", "/"), "/")
	s.mu.RLock()
	defer s.mu.RUnlock()
	current := ""
	for _, part := range strings.Split(normalized, "/") {
		if part == "" {
			continue
		}
		if current == "" {
			current = part
		} else {
			current += "/" + part
		}
		perm, ok := s.entries[current]
		if !ok {
			continue
		}
		users, ok := perm[action]
		if !ok {
			continue
		}
		granted := false
		for _, u := range users {
			if u == "*" || u == user {
				granted = true
				break
			}
		}
		if !granted {
			return false
		}
	}
	return true
}

// Set mirrors set_folder_permission (validated, merged over defaults).
func (s *Store) Set(path, action string, users []string) error {
	normalizedPath, err := NormalizePermissionPath(s.root, path)
	if err != nil {
		return err
	}
	if !validAction(action) {
		return errors.New("invalid permission action")
	}
	nu, err := NormalizePermissionUsers(users)
	if err != nil {
		return err
	}
	s.mu.Lock()
	current := map[string][]string{}
	for k, v := range DefaultPermission {
		current[k] = append([]string{}, v...)
	}
	for k, v := range s.entries[normalizedPath] {
		current[k] = append([]string{}, v...)
	}
	current[action] = nu
	s.entries[normalizedPath] = current
	s.mu.Unlock()
	return s.Save()
}

// Delete mirrors delete_folder_permission.
func (s *Store) Delete(path string) bool {
	normalizedPath, err := NormalizePermissionPath(s.root, path)
	if err != nil {
		return false
	}
	s.mu.Lock()
	if _, ok := s.entries[normalizedPath]; !ok {
		s.mu.Unlock()
		return false
	}
	delete(s.entries, normalizedPath)
	s.mu.Unlock()
	_ = s.Save()
	return true
}

// Save persists validated entries atomically (temp + rename).
func (s *Store) Save() error {
	s.mu.RLock()
	snapshot := map[string]map[string][]string{}
	for path, entry := range s.entries {
		np, ne, err := NormalizePermissionEntry(s.root, path, entry)
		if err != nil {
			continue
		}
		snapshot[np] = ne
	}
	s.mu.RUnlock()
	out, err := json.MarshalIndent(snapshot, "", "    ")
	if err != nil {
		return err
	}
	dest := filepath.Join(s.root, PermissionsFile)
	tmp, err := os.CreateTemp(s.root, ".webshare_perm_*.tmp")
	if err != nil {
		return err
	}
	tmpName := tmp.Name()
	defer os.Remove(tmpName)
	if _, err := tmp.Write(append(out, '\n')); err != nil {
		tmp.Close()
		return err
	}
	if err := tmp.Close(); err != nil {
		return err
	}
	return os.Rename(tmpName, dest)
}

// Load reads the store file, skipping invalid legacy entries.
func (s *Store) Load() error {
	data, err := os.ReadFile(filepath.Join(s.root, PermissionsFile))
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}
	var loaded map[string]map[string][]string
	if err := json.Unmarshal(data, &loaded); err != nil {
		return err
	}
	normalized := map[string]map[string][]string{}
	for path, entry := range loaded {
		np, ne, err := NormalizePermissionEntry(s.root, path, entry)
		if err != nil {
			continue
		}
		normalized[np] = ne
	}
	s.mu.Lock()
	s.entries = normalized
	s.mu.Unlock()
	return nil
}

// RoleCanMutate mirrors role_can_mutate: admins always, guests only when
// guest upload is enabled.
func RoleCanMutate(role string, allowGuestUpload bool) bool {
	if role == "admin" {
		return true
	}
	return allowGuestUpload
}

// EnsureMutationAllowed mirrors ensure_mutation_allowed.
func EnsureMutationAllowed(role string, allowGuestUpload bool) (bool, string, int) {
	if RoleCanMutate(role, allowGuestUpload) {
		return true, "", 200
	}
	return false, "업로드/변경 권한이 없습니다", 403
}

// EnsurePathAccess mirrors ensure_path_access: protected check first,
// then the permission engine. Returns (ok, message, status).
func (s *Store) EnsurePathAccess(path, action, role string) (bool, string, int) {
	normalized := NormalizeRelativePath(path)
	if IsProtectedSystemPath(normalized) {
		return false, "시스템 경로 접근이 차단되었습니다", 403
	}
	if !s.Check(normalized, role, action) {
		return false, "권한이 없습니다", 403
	}
	return true, "", 200
}

// BuildCapabilities mirrors build_path_capabilities.
func (s *Store) BuildCapabilities(path, role string, isDir bool, itemType string, allowGuestUpload bool) Capabilities {
	normalized := NormalizeRelativePath(path)
	parent := GetParentRelativePath(normalized)
	mut := RoleCanMutate(role, allowGuestUpload)
	read := !IsProtectedSystemPath(normalized) && s.Check(normalized, role, "read")
	write := mut && !IsProtectedSystemPath(normalized) && s.Check(normalized, role, "write")
	del := mut && !IsProtectedSystemPath(normalized) && s.Check(normalized, role, "delete")
	parentWrite := mut && !IsProtectedSystemPath(parent) && s.Check(parent, role, "write")
	uploadMkdir := write
	if !isDir {
		uploadMkdir = parentWrite
	}
	return Capabilities{
		Read:   read,
		Write:  write,
		Delete: del,
		Rename: del && parentWrite,
		Move:   del,
		Copy:   mut && read,
		Upload: uploadMkdir,
		Mkdir:  uploadMkdir,
		Edit:   write && !isDir,
		Trash:  del,
		Unzip:  mut && !isDir && itemType == "archive" && read && parentWrite,
	}
}
